### PR TITLE
[cheerio] Refactor using namespace to expose internal types for callers

### PR DESCRIPTION
**Pull request by paulmelnikow**
This makes it possible to write a function like this, which accepts these internal types:

```ts
function hasSelector(parsed: CheerioSelector, selector: string) {
  return parsed(selector).length
}
```

Rename `CheerioStatic` to ~~`CheerioRoot`~~ `Root`. Objects of this type have static methods which mirror static methods of `cheerio`, however the objects themselves aren't static, so having "static" in the type name is confusing. Also rename `CheerioElement` to `Element` and `CheerioSelector` to `Selector` at upstream author's request.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
    - I couldn't get this to run cleanly, though the failure is not related to this change.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: N/A
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by paulmelnikow** [comment]
I'm seeing this error in the CI logs:

```
##[error]../cheerio/index.d.ts(303,5): error TS2666: Exports and export assignments are not permitted in module augmentations.
```

but I'm not sure how to fix it. Any suggestions?

**Comment by paulmelnikow** [comment]
Is the problem from using `export` and `declare module` together? Is there a way I can do this without `declare module` that would work better?

**Comment by privatenumber** [comment]
I'm still learning a lot of this as I go but I will try to help.

A lot of the popular modules I see in this codebase use `namespace`:
- [Webpack](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/webpack/index.d.ts)
- [acorn](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/acorn/index.d.ts)
- [react](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L66)

Which makes me wonder if cheerio should be using `namespace` instead, but based on [this doc](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html), `module` is  recommended over `namespace` and shows you how to import the type from a module (which I guess is what you're trying to achieve).

I did a quick search to see who's using `module` and how they're using it, which might offer some direction forward:
- [rx-core](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rx-core/index.d.ts) uses a combination of `namespace` and `module`
- Node [path](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/events.d.ts) also uses a combination of `namespace` and `module
- Node [event](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/events.d.ts)

Hope that helps as reference

**Comment by paulmelnikow** [comment]
Hey @privatenumber thanks for those links! I reworked a version that combined namespace and module and got it working. It also allows me to use e.g. `cheerio.CheerioSeletor` which is what I needed to do. It's passing lint locally; will see how it fares in the cloud.

**Comment by paulmelnikow** [comment]
There's a lint failure in `types/enzyme/index.d.ts`. It doesn't make a lot of sense to me. (It goes away if I remove the occurrence of `@deprecated`.)

```
The 'no-redundant-jsdoc-2' rule threw an error in '/home/runner/work/DefinitelyTyped/DefinitelyTyped/types/enzyme/index.d.ts':
Error: Unexpected tag kind: JSDocDeprecatedTag
    at checkTag (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:118:23)
    at cb (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:67:25)
    at visitNodes (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/typescript/lib/typescript.js:26992:30)
    at Object.forEachChild (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/typescript/lib/typescript.js:27310:21)
    at cb (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:72:19)
    at Array.forEach (<anonymous>)
    at walk (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:56:34)
    at Rule.AbstractRule.applyWithFunction (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/tslint/lib/language/rule/abstractRule.js:39:9)
    at Rule.apply (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/rules/noRedundantJsdoc2Rule.js:36:21)
    at Linter.applyRule (/home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/tslint/lib/linter.js:214:29)
```

**Review by ljharb — APPROVED** [review]
chai-enzyme and enzyme changes LGTM

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## cheerio ([<kbd>unpkg</kbd>](https://unpkg.com/browse/cheerio@latest/))
was missing the following properties:
1. version
## enzyme-to-json ([<kbd>unpkg</kbd>](https://unpkg.com/browse/enzyme-to-json@latest/))
was missing the following properties:
1. renderToJson
1. shallowToJson
1. mountToJson
## json_ml ([<kbd>unpkg</kbd>](https://unpkg.com/browse/json_ml@latest/))
was missing the following properties:
1. stringifyListMode
1. stringifyElement
1. parseElementList
1. parseElement
1. JsonML
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 11e8c148cc312acb7682504d692d9f82ac55197d
</p>

File: types/enzyme/tslint.json
Comment (paulmelnikow) [inline types/enzyme/tslint.json:7]: See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46006#issuecomment-658956797

**Comment by paulmelnikow** [comment]
> ## cheerio ([unpkg](https://unpkg.com/browse/cheerio@latest/))
> 
> was missing the following properties:
> 
>     1. version

I fixed this.

> ## enzyme-to-json ([unpkg](https://unpkg.com/browse/enzyme-to-json@latest/))
> 
> was missing the following properties:
> 
>     1. renderToJson
> 
>     2. shallowToJson
> 
>     3. mountToJson
> 
> 
> ## json_ml ([unpkg](https://unpkg.com/browse/json_ml@latest/))
> 
> was missing the following properties:
> 
>     1. stringifyListMode
> 
>     2. stringifyElement
> 
>     3. parseElementList
> 
>     4. parseElement
> 
>     5. JsonML

I did not investigate nor fix these.

**Review by ljharb — APPROVED** [review]

File: types/enzyme/tslint.json
Comment (sheetalkamat) [reply types/enzyme/tslint.json:7]: Do update the lint file. merge with master and things should work out. @orta is fixing this kind of issue.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/enzyme/tslint.json
Comment (paulmelnikow) [reply types/enzyme/tslint.json:7]: Do you mean _don't_ update the lint config? i.e. I should revert this, merge master, and this will disappear?

Comment (paulmelnikow) [reply types/enzyme/tslint.json:7]: @sheetalkamat Could you clarify?

**Comment by paulmelnikow** [comment]
@sheetalkamat Hi, did you see my question at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46006#discussion_r466151182? This has been open for over a month and I'd love to get it merged.

**Comment by paulmelnikow** [comment]
I’m waiting for a response to a question. Please don’t close the PR. It needs maintainer attention. Thanks.

File: types/cheerio/index.d.ts
Comment (matthewmueller) [inline types/cheerio/index.d.ts:19]: Could we just call this `Element`? `cheerio.Element` looks a bit nicer than `cheerio.CheerioElement` :-)

File: types/cheerio/index.d.ts
Comment (matthewmueller) [inline types/cheerio/index.d.ts:270]: Could we just call this `Selector`? `cheerio.Selector` looks a bit nicer than `cheerio.CheerioSelector` :-)

File: types/cheerio/index.d.ts
Comment (matthewmueller) [inline types/cheerio/index.d.ts:283]: Could we just call this `Root`? `cheerio.Root` looks a bit nicer than `cheerio.Root` :-)

File: types/enzyme/tslint.json
Comment (paulmelnikow) [reply types/enzyme/tslint.json:7]: I'm going to try reverting it and seeing if it's working now.

File: types/cheerio/index.d.ts
Comment (paulmelnikow) [reply types/cheerio/index.d.ts]: Updated!

File: types/cheerio/index.d.ts
Comment (paulmelnikow) [reply types/cheerio/index.d.ts]: Updated!

File: types/cheerio/index.d.ts
Comment (paulmelnikow) [reply types/cheerio/index.d.ts]: Updated!

Comment (ljharb) [reply types/cheerio/index.d.ts]: Users tho won't be seeing the `cheerio.`, they'll be importing `Element`, which would likely be confusing.

Comment (paulmelnikow) [reply types/cheerio/index.d.ts]: Personally I agree that CheerioElement is better, because there are a lot of kinds of Elements. Although nothing stops developers from importing the namespace just as before.

Comment (matthewmueller) [reply types/cheerio/index.d.ts]: Good points! My usage pattern seems different than yours and it might not be the recommended approach.

I expected usage to look like this:

```ts
const cheerio = require('cheerio')
// or 
import cheerio from 'cheerio'

function toHTML(element: cheerio.Element): string { ... }
```

Would this work?

Comment (ljharb) [reply types/cheerio/index.d.ts]: I think that what's _far_ more common, and what vscode's automatic importing would do for the type, is:
```jsx
import cheerio, { Element } from 'cheerio';

const $ = cheerio.load("...")

function toHTML(element: Element): string { ... }
```

Comment (paulmelnikow) [reply types/cheerio/index.d.ts]: This got merged, though I'm happy to revisit this in a follow-on. The plan is to move the types upstream (cheeriojs/cheerio#1471) so perhaps the follow-on changes can happen there.

**Comment by solarmosaic-kflorence** [comment]
FYI this broke our builds, which had not been using a pinned version of the `@types/cheerio` dependency. To fix we had to pin it to `0.22.21` -- which indicates this may not have been a backwards compatible change, in which case, it would be nice not to publish it under a patch version update.

**Comment by pluma** [comment]
Yeah, I also had fun with this via `json_ml`.

**Comment by paulmelnikow** [comment]
Hmm, AFAIK DT does not promise backward compatibility of types through patch releases. Not sure if there is a solution for making sure compatible versions of types' dependencies are installed…

**Comment by solarmosaic-kflorence** [comment]
I would question the usage of semantic versioning if the semantics are not respected... anyways, like most other frontend dependencies we are locking this one in since we can't trust publishers to be backwards compatible outside of major versions.

**Comment by paulmelnikow** [comment]
Pinning them is a good idea. See DefinitelyTyped/DefinitelyTyped#13225 for some discussion about what the version numbers mean in DT.

**Comment by ljharb** [comment]
@solarmosaic-kflorence DT explicitly does not use semantic versioning, and DT package versions should always be pinned.

**Comment by astorije** [comment]
I opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50133 to pin `@types/cheerio` as we're seeing this locally in all our repos until we manually futz with the `yarn.lock` files:

```
node_modules/@types/enzyme/index.d.ts:384:15 - error TS2503: Cannot find namespace 'cheerio'.

384     render(): cheerio.Cheerio;
                  ~~~~~~~

node_modules/@types/enzyme/index.d.ts:728:69 - error TS2503: Cannot find namespace 'cheerio'.

728 export function render<P, S>(node: ReactElement<P>, options?: any): cheerio.Cheerio;
                                                                        ~~~~~~~


Found 2 errors.
```

**Pull request merged**
